### ISSUE TITLE
EVM Proposal: userspace selfdestruct

### ISSUE THREAD

**Issue by Stebalien** [filecoin-project/ref-fvm#1221]
Our current SELFDESTRUCT operation is incorrect because it deletes the contract immediately. Also, see #1174.

Proposal: Implement SELFDESTRUCT in userspace by:

1. Sending all funds to the beneficiary.
2. Marking the contract as "selfdestructed" in the contract's actor state by recording the current origin and nonce.
3. Changing the EVM actor to "pretend like it has no code" if it (a) sees this field set and (b) the nonce/origin _don't_ match (i.e., we've moved on to a new transaction).
4. Adding a new `Resurrect` function that behaves like the `Constructor` but:
    1. Must be called directly by the EAM.
    2. Can only be called if the selfdestruct field is set.
    3. Blows away state and re-constructs the EVM contract.
5. Change the EAM to call `Resurrect` when a user attempts to call `CREATE2` and the target actor already exists.

Changes WRT Ethereum:

1. This won't ever delete state and/or refunds. But, whatever.
2. Unlike Ethereum, funds sent to a contract after SELFDESTRUCT but before the end of the transaction won't poof into the void and will isntead remain with the contract. But I consider that a feature.

**Comment by vyzo** [comment]
SGTM.
We want it when?

**Comment by raulk** [comment]
This sounds reasonable, especially given that the Ethereum community has noted the many quirks of the SELFDESTRUCT opcode and is potentially looking to [eliminate it](https://hackmd.io/@vbuterin/selfdestruct). Before that happens, it's a good call to keep those quirks contained in the EVM runtime actor. Just one question: could you elaborate on the reason why we don't delink the storage HAMT immediately? Can we re-enter within the same transaction with the expectation that storage will still be alive?

**Comment by Stebalien** [comment]
> Just one question: could you elaborate on the reason why we don't delink the storage HAMT immediately? Can we re-enter within the same transaction with the expectation that storage will still be alive?

Exactly this. Deletion is deferred until the end of the transaction.

**Comment by Stebalien** [comment]
Current status: the builtin actors use the latest SDK so this is unblocked, but I haven't made any progress towards actually implementing it. You're good to go vyzo!

**Comment by mriise** [comment]
Please be aware of behavior of EIP-161 (and if we choose to adopt it, EIP-4747), since there are some cases where there is an auto deletion of "empty" accounts.

We will still be doing userspace deletes for these as well though.

cc @vyzo

**Comment by raulk** [comment]
@Stebalien in your refactor of SELFDESTRUCT, can you perform the audit specified here: https://github.com/filecoin-project/ref-fvm/issues/1221.

**Comment by Stebalien** [comment]
We discussed bypassing the EAM on resurrect, but after writing this all out, I'm nixing that approach as it would require computing the expected EVM address in multiple places.

**Comment by Stebalien** [comment]
This also requires changing the EAM's `Create2` and `Create` signatures to return an `Option<Address>` for the "predictable" address, because they may be "resurrecting" an actor.
